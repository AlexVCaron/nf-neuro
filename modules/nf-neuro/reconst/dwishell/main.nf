process RECONST_DWISHELL {
    tag "$meta.id"
    label 'process_single'

    container "scilus/scilpy:2.2.0_cpu"

    input:
        tuple val(meta), path(dwi), path(bval), path(bvec)
        val(shell_to_fit)

    output:
        tuple val(meta), path("*__dwi_sh_shells.nii.gz"), path("*__bval_sh_shells"), path("*__bvec_sh_shells"), emit: dwi_shells
        tuple val(meta), path("*__out_indices"), emit: out_indices, optional: true
        path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"

    def out_indices = task.ext.out_indices ? "--out_indices ${prefix}__out_indices" : ''
    def block_size = task.ext.block_size ? "--block_size ${task.ext.block_size}" : ''
    def tolerance = task.ext.tolerance ? "--tolerance ${task.ext.tolerance}" : ''
    def force = task.ext.force ? "-f" : ''

    """
    scil_dwi_extract_shell $dwi $bval $bvec $shell_to_fit \
        ${prefix}__dwi_sh_shells.nii.gz \
        ${prefix}__bval_sh_shells \
        ${prefix}__bvec_sh_shells \
        ${out_indices} \
        ${block_size} \
        ${tolerance} \
        ${force}

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """

    stub:
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    echo $args

    touch ${prefix}__dwi_sh_shells.nii.gz
    touch ${prefix}__bval_sh_shells
    touch ${prefix}__bvec_sh_shells

    if ${task.ext.out_indices} {
        touch ${meta.id}__out_indices
    }

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: \$(uv pip -q -n list | grep scilpy | tr -s ' ' | cut -d' ' -f2)
    END_VERSIONS
    """
}
