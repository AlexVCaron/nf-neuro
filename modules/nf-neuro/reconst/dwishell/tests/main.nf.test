// nf-core modules test reconst/dwishell
nextflow_process {

    name "Test Process RECONST_DWISHELL"
    script "../main.nf"
    process "RECONST_DWISHELL"
    config "./nextflow.config"

    tag "modules"
    tag "modules_nfneuro"
    tag "reconst"
    tag "reconst/dwishell"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
        run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
            script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
            process {
                """
                input[0] = Channel.from( [ "DWIss1000-dir32.zip" ] )
                input[1] = "test.load-test-data"
                """
            }
        }
    }

    test("dwishell") {

        when {
            process {
            """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]}
            """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                    process.out.dwi_shells,
                    process.out.versions).match() }
            )
        }
    }

    test("dwishell - stub") {

        tag "stub"
        options "-stub-run"

        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/dwi.nii.gz"),
                        file("\${test_data_directory}/dwi.bval"),
                        file("\${test_data_directory}/dwi.bvec")
                    ]}
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.versions).match() }
            )
        }

    }
}
