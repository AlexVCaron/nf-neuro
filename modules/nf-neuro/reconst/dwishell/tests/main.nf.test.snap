{
    "dwishell - stub": {
        "content": [
            [
                "versions.yml:md5,a77958e6a1f7e59f33a6efac6e39eb77"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-01T18:20:22.139824591"
    },
    "dwishell": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__dwi_sh_shells.nii.gz:md5,0e14081d7cc15740169186fb7df8ec20",
                    "test__bval_sh_shells:md5,5e0381be0d5c8b3bb8bcd426d0d9cc16",
                    "test__bvec_sh_shells:md5,529513ab12df505acd472fe8ebd2b94f"
                ]
            ],
            [
                "versions.yml:md5,a77958e6a1f7e59f33a6efac6e39eb77"
            ]
        ],
        "meta": {
            "nf-test": "0.9.0",
            "nextflow": "25.04.7"
        },
        "timestamp": "2025-10-01T18:20:18.474023782"
    }
}